Aijuh/74/implement-log-viewer

Aijuh opened the pull request:
Implemented logger app by using same code as qiwis' logger.py.
An image below is implemented logger app. We can see RID value message from builder.py. Position and log format should be determined.

Unfortunately, other app also using logging so unwanted log message printed out since we are using root logger.
I think we can handle it by making naming convention inside iquip app. If other app running by iquip's app wants to print log message, we may read it  inside that specific iquip's app. 
I didn't think all the details but give option (ex. bool variable) to the Baseapp to read all log message or not. If it is supposed to take all the log message, read it using root logger and send to its own logger.
I may missed some points. I will wait for further comments.
 This closes issue #74

![image](https://github.com/snu-quiqcl/iquip/assets/43592775/c8490210-3a3e-458f-aa9d-63e89a383a86)

Aijuh committed:
Add logger app

Same code as qiwis's logger.py.

Change all apps' print to logging.

Aijuh committed:
line_too_long error

kangz12345 commented:
> Unfortunately, other app also using logging so unwanted log message printed out since we are using root logger.

Good! I intended that. That's why we used the root logger.
We can use the filter options for the logger of the handler to filtering the logs.
You can implement that feature later as well!

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,202 @@
+        formatter = logging.Formatter(fs)
+        self.handler.setFormatter(formatter)
+        logger = logging.getLogger()
+        logger.setLevel(logging.DEBUG)
```
In my opinion, it is more efficient to change the level when the user select the level via the combobox.
For example, when the selected level is `WARNING`, many useless log records will be passed to the root logger, which is quite inefficient.

Aijuh replied:
I will set root logger's level as `WARNING` and change the code of level selector.

Aijuh committed:
Delete receivedSlot func&set initial log level as WARNING

1. Delete receivedSlot function in LoggerApp because it is not used.

2. Instead of changing directly changing level of root logger inside logger.py, it remains to manipulate handler's level.
But initial level of handler's level is changed to warning. So it didn't print any minor message at the start point.
In order to inform users that current log level is "WARNING", I used ComboBox.setCurentText method to change current displayed text.

Aijuh committed:
Add Filter

1. (line 207) Change boradcast to logger.

2. (line 159) Give detailed log format.

3. Add Filter to pass specific logger name.

Aijuh commented:
I finished setting initial log handler's level as `WARNING`. Initial GUI window is shown as below.

![image](https://github.com/snu-quiqcl/iquip/assets/43592775/c8367b3a-9ceb-4e78-8ff4-f3d4aa3e674a)

Aijuh commented:
Also I add a filter inside log handler to pass only designated name of logger.
In the below image, even if the level is `DEBUG` only log message from iquip app printed.
For now, allowed log channel is : `["iquip.apps.builder", "iquip.apps.explorer","iquip.apps.logger", "iquip.apps.thread"]`
![image](https://github.com/snu-quiqcl/iquip/assets/43592775/d01eea92-0d57-44c1-a3a2-368600dcf2b6)

Aijuh commented:
I remember that I didn't get exact feedbacks about log filter.
Is it okay to remove the filter part in this PR and just leave log viewer part? If that is the case, this PR will be only about log viewer.
I think we should think more about filter option. I will add option to write log to the text file in next PR.

kangz12345 commented:
> I remember that I didn't get exact feedbacks about log filter.
> 
> Is it okay to remove the filter part in this PR and just leave log viewer part? If that is the case, this PR will be only about log viewer.
> 
> I think we should think more about filter option. I will add option to write log to the text file in next PR.


I'm sorry for the late reply.
Yes, I also think that the filter should be implemented as another feature.
Please remove the filter in this PR and design and implement the filter feature in another PR.

Aijuh committed:
Delete filter part

Aijuh commented:
Remove the filter part of the code.

kangz12345 commented on iquip/apps/builder.py:
```diff
@@ -260,7 +264,7 @@ def run(self):
             rid = response.json()
         except requests.exceptions.RequestException as err:
-            print(err)
+            logger.exception(err)
```
`logger.exception("Failed to submit the experiment.")`

BECATRUE replied:
You should print more detailed description in terms of what exception occurs. (Like Jiyong's suggestion!)

Previously, I just printed `err` because the logger was not implemented yet😅

kangz12345 commented on iquip/apps/explorer.py:
```diff
@@ -84,7 +88,7 @@ def run(self):
             experimentList = response.json()
         except requests.exceptions.RequestException as err:
-            print(err)
+            logger.exception(err)
```
`logger.exception("Failed to fetch the file list.")`

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+
+
+class ConfirmClearingFrame(QWidget):
+    """
+    A confirmation frame for log clearing.
+    """
```
You can write it in a single line.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+    """
+    A confirmation frame for log clearing.
+    """
+    confirmed = pyqtSignal()
```
1 blank line before this.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+    confirmed = pyqtSignal()
+
+    def __init__(self, parent: Optional[QObject] = None):
+        """
+        Extended.
+        """
```
Please write it in a single line.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        """
+        super().__init__(parent=parent)
+        # widgets
+        self.label = QLabel("Are you sure to clear?")
+        self.buttonBox = QDialogButtonBox()
```
Please set the parents: `QLabel("...", self)`.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        # widgets
+        self.logEdit = QTextEdit(self)
+        self.logEdit.setReadOnly(True)
+        self.clearButton = QPushButton("Clear")
```
`QPushButton("Clear", self)`

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        self.buttonBox.accepted.connect(self.buttonOKClicked)
+        self.buttonBox.rejected.connect(self.buttonCancelClicked)
+        # layouts
+        layout = QVBoxLayout()
+        self.setLayout(layout)
```
You can write `layout = QVBoxLayout(self)`.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        layout.addWidget(self.buttonBox)
+
+    def buttonOKClicked(self):
+        """Clicks OK to clear log."""
```
It seems like this method is clicking the OK button.
How about "Called when the OK button is clicked."?

Aijuh replied:
It seems reasonable. I will also change same part in qiwis.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+
+    Manages a logger frame.
+
+    Attributes:
```
It has more attributes: `confirmFrame` and `handler`.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        self.loggerFrame.levelBox.setCurrentText("WARNING")
+
+    @pyqtSlot(str)
+    def setLevel(self, text: str):
```
Can you change the root logger's level here as well?

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+
+        Args:
+            text: Selected level in the level select box.
+                  It should be one of "DEBUG", "INFO", "WARNING", "ERROR" and "CRITICAL".
+                  It should be case-sensitive and any other input is ignored.
```
We put only 2 more spaces:
```
Args:
    text: ...
      It should be ...
      It should be ...
```

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+    @pyqtSlot()
+    def checkToClear(self):
+        """Shows a confirmation frame for log clearing."""
+        logger.info("Clicked to clear logs")
```
Please be more specific: "Clear button is clicked to clear logs".

kangz12345 commented on iquip/apps/thread.py:
```diff
@@ -54,7 +58,7 @@ def run(self):
             data = response.json()
         except requests.exceptions.RequestException as err:
-            print(err)
+            logger.exception(err)
```
`logger.exception("Failed to fetch the experiment information.")`

kangz12345 commented on iquip/apps/thread.py:
```diff
@@ -65,4 +69,4 @@ def run(self):
             )
         else:
-            print("The selected item is a non-experiment file.")
+            logger.error("The selected item is a non-experiment file.")
```
Although it is the original text, but I think "The selected item is not an experiment file." sounds better.

BECATRUE replied:
How about modifying the level as `logging.info()`?
I think the case where the selected item is not an experiment file is not a critical problem.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -352,7 +356,7 @@ def onSubmitted(self, rid: int):
         TODO(BECATRUE): It will be developed in Log Viewer project.
         """
-        print(f"RID: {rid}")
+        logger.info("RID: %d", rid)
```
How about updating the method docstring as follows:
```
Sends the rid to the logger after submitted.
```

This is just an example, and if you think 'prints' is still ok, never mind!

BECATRUE replied:
Also, please remove the above TODO comment because you implemented it instead of me.

Aijuh committed:
Updated string format

Aijuh committed:
Update logger.py

Aijuh commented:
Updated requested changes.

BECATRUE commented:
After merging this PR, please implement the unit tests for `logger.py`!

Aijuh commented:
I don't know much about Unittest. I will read previous issues and PR and will implement it.

BECATRUE commented:
> I don't know much about Unittest. I will read previous issues and PR.

Okay! Actually, I hoped you will learn about Python unittest through this.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -76,18 +76,15 @@ def __init__(self, parent: Optional[QObject] = None):
         # widgets
-        self.label = QLabel("Are you sure to clear?")
+        self.label = QLabel("Are you sure to clear?", self)
         self.buttonBox = QDialogButtonBox()
```
Add a parent to this as well!

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -101,12 +98,12 @@ def __init__(self, parent: Optional[QObject] = None):
 
     def buttonCancelClicked(self):
-        """Clicks Cancel not to clear log."""
+        """Clicked Cancel not to clear log."""
```
For consistency, please follow the form of `buttonOKClicked()` docstring.
"Called when the Cancel button is clicked."

Aijuh replied:
I had fixed it in this way, but since it is not requested part, I just moved to the original one.
I will change it.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -117,6 +114,8 @@ class LoggerApp(qiwis.BaseApp):
 
     Attributes:
         loggerFrame: A frame that shows the logs.
+        confirmFame: A frame that asks to whether clear logs.
```
"... whether to clear logs" seems better.

Aijuh committed:
Updated feedback from @BECATRUE

Aijuh committed:
Update feedback from @kangz12345

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -85,14 +85,14 @@ def __init__(self, parent: Optional[QObject] = None):
         # layouts
-        layout = QVBoxLayout()
+        layout = QVBoxLayout(self)
         self.setLayout(layout)
```
You can remove this line since you set the parent of `layout`.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -114,7 +114,7 @@ class LoggerApp(qiwis.BaseApp):
     Attributes:
         loggerFrame: A frame that shows the logs.
-        confirmFame: A frame that asks to whether clear logs.
+        confirmFame: A frame that asks to whether to clear logs.
```
I'm sorry for the confusion, "... asks whether to clear logs."

BECATRUE replied:
The `confirmFame` looks a typo!

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        }
+        if text in level:
+            self.handler.setLevel(level[text])
+            (logging.getLogger()).setLevel(level[text])
```
You don't need the parenthesis: `logging.getLogger().setLevel(...)`

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+class LoggerApp(qiwis.BaseApp):
+    """App for logging.
+
+    Manages a logger frame.
```
Please write more explanation. It looks too vague.

BECATRUE replied:
For example, there is another frame, confirmFrame!

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+    Attributes:
+        loggerFrame: A frame that shows the logs.
+        confirmFame: A frame that asks to whether to clear logs.
+        handler: A handler for adding logs to GUI. 
```
How about "A handler to add logs to the loggerFrame`?

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        """Extended.
+
+        Args:
+            name: Name of the App
```
Please remove this explanation because it is already written in `BaseApp` docstring!
"Extended" expression implies that arguments not shown here are already described in the super class.

Aijuh committed:
feedback from @kangz12345

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        self.loggerFrame = LoggerFrame()
+        # connect signals to slots
+        self.loggerFrame.clearButton.clicked.connect(self.checkToClear)
+        self.confirmFrame = ConfirmClearingFrame()
```
Please move this up to the above comment.
Several comments look necessary to show legibly.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        self.confirmFrame.confirmed.connect(self.clearLog)
+        self.handler = LoggingHandler(self.addLog)
+        # TODO(aijuh): Change the log format when it is determined.
+        fs ="%(levelname)s [%(name)s] [%(filename)s:%(lineno)d] %(message)s "
```
The variable name looks vague, but it will be updated later, so you can ignore it now.

Aijuh replied:
I will make two handlers and two formats when implement option to add log to file. I will fix name at that PR.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,189 @@
+        
+        Emits input signal to the connected function.
+        """
+        s = self.format(record)
```
I think we try not to use a single character name.
According to this [styleguide](https://google.github.io/styleguide/pyguide.html#3161-names-to-avoid), it is not recommended.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+        self.loggerFrame.levelBox.setCurrentText("WARNING")
+
+    @pyqtSlot(str)
+    def setLevel(self, text: str):
```
How about `levelText`? `text` looks ambiguous.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+        }
+        if text in level:
+            self.handler.setLevel(level[text])
+            logging.getLogger().setLevel(level[text])
```
@kangz12345 Did you intend this?

If so, in `__init__()`, the root logger level and the handler level are set differently, and it looks too awkward.

BECATRUE replied:
Also, in `__init__()`, you can just call `self.setLevel("WARNING")`.

kangz12345 replied:
1. The root logger level
2. The handler level
3. The combobox level text

should always coincide.

BECATRUE replied:
Okay! I think it is okay to remain the root looger level as the lowest level, i.e. DEBUG, but keeping them all the same looks good, too.

kangz12345 replied:
I think they should _always_ be the same, so please unify those levels in `__init__()` as well.

Aijuh replied:
I intended to prevent possible problems by changing root logger's level.
 But as I know there is no problem with handling root logger's level.

 I will change this part.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+        Args:
+            content: Received log message.
+        """
+        timeString = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(time.time()))
```
It is just an option, and you can use `QDateTime.currentDateTime()` to obtain the current time.
I used this in `builder.py` and it may look more unified.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+
+    @pyqtSlot()
+    def checkToClear(self):
+        """Shows a confirmation frame for log clearing."""
```
Isn't "... for clearing logs" more right?

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+    @pyqtSlot()
+    def checkToClear(self):
+        """Shows a confirmation frame for log clearing."""
+        logger.info("Clear button is clicked to clear logs")
```
The statement is exactly correct, but it looks a little confused because "clear" is repeated.

How about "Tried to clear logs"?

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+"""Module for log viewer in apps."""
```
How about "App module for showing logs."?

Aijuh committed:
feedback from @BECATRUE

Aijuh committed:
time string update using QDatetime

Aijuh committed:
_init_:logLevel is warning

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+class LoggingHandler(logging.Handler):
+    """Handler for logger.
+
+    Sends a log message to connected function using emit.   
```
'... to the connected'

BECATRUE replied:
Also, 'through a signal' looks better that 'using emit`.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+
+
+class ConfirmClearingFrame(QWidget):
+    """A confirmation frame for log clearing."""
```
I realized just now that there is no any docstring about its attributes!

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+    """App for logging.
+
+    Set a handler of root logger and manages loggerFrame to show log messages.
+    Give options to clear logs and select log level in the loggerFrame.
```
'Sets' and 'Gives'.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+        rootLogger = logging.getLogger()
+        rootLogger.setLevel("WARNING")
+        rootLogger.addHandler(self.handler)
+        self.handler.setLevel("WARNING")
```
Please use `self.setLevel("WARNING")`.

BECATRUE replied:
And does `self.loggerFrame.levelBox.setCurrentText("WARNING")` call `self.setLevel("WARNING")`?
If so, I think you can just remove these two lines.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+        """Responds to the setLevelBox widget and changes the handler's level.
+
+        Args:
+            text: Selected level in the level select box.
```
Please update the argument name.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+              It should be one of "DEBUG", "INFO", "WARNING", "ERROR" and "CRITICAL".
+              It should be case-sensitive and any other input is ignored.
+        """
+        level = {
```
`levels` looks more appropriate because it is not only one level.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+        Args:
+            content: Received log message.
+        """
+        timeString = QDateTime.currentDateTime().toString("yyyy-MM-dd HH:mm:ss")
```
👍

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,184 @@
+
+    @pyqtSlot(str)
+    def setLevel(self, levelText: str):
+        """Responds to the setLevelBox widget and changes the handler's level.
```
Isn't the levelBox correct?

Aijuh replied:
My bad. Next time I change variable name, I will change all the comments using that name.

BECATRUE reviewed (changes requested):
Maybe almost done!

Aijuh committed:
feedback from @BECATRUE

Aijuh commented:
I updated codes and got a question. 
I will implement feature for saving logs to file for next PR.
Before, I set root logger as "DEBUG" and change two different handlers' level. (ex. set handler for the loggerFrame as "WARNING", handler for the filesaving as "INFO").
But since level of root logger and LoggerFrame handler change at same time,  file saving handler's level may be ignored.
So which way is better?

1. Two handlers and root logger has all same level.
2.  Set root logger as DEBUG and control logs by handler's level.

kangz12345 commented:
Good question.
In that case:
1. GUI handler level = GUI handler level combobox
2. File handler level = File handler level combobox
3. Root logger level = minimum of 1 and 2 (note that [logger levels](https://docs.python.org/3/library/logging.html#logging-levels) have numeric values so you can use `min()` to get the lower level)

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+            slotfunc: A slot function which is called when a log record is emitted.
+        """
+        super().__init__()
+        self.signaller = _Signaller()
```
Please write down the docstring about `signaller` attribute.

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+    
+    Attributes:
+        label: Displays a confirmation message to clear logs in the LoggerFrame.
+        buttonBox: Contains OK and Cancel button to check whether to clear logs.
```
Please write in the format:
```
The label for displaying ...
The button box with OK and ...
```

BECATRUE commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,188 @@
+        buttonBox: Contains OK and Cancel button to check whether to clear logs.
+    """
+
+    confirmed = pyqtSignal()
```
Please add the docstring about this signal.

kangz12345 replied:
You should add "Signals:" section in the docstring.

BECATRUE reviewed (changes requested):
This may be the final review!

After done, please request Jiyong for a review.

Aijuh committed:
feedback from @BECATRUE

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -0,0 +1,192 @@
+        self.confirmFrame.confirmed.connect(self.clearLog)
+        self.handler = LoggingHandler(self.addLog)
+        # TODO(aijuh): Change the log format when it is determined.
+        fs ="%(levelname)s [%(name)s] [%(filename)s:%(lineno)d] %(message)s "
```
Oh, why isn't there a space between `=` and `"`, and why is there a space at the end?
In other words, shouldn't it be `fs = "%(levelname)s ... %(message)s"`?

Aijuh replied:
My mistake. I will check details.

Aijuh committed:
Update comments

Aijuh commented:
Fix comments.

kangz12345 commented on iquip/apps/logger.py:
```diff
@@ -172,21 +174,21 @@ def frames(self) -> Tuple[LoggerFrame]:
 
         Args:
-            content: Received log message.
+            content: A received log message.
```
You don't need "A" here since it is just a noun and describes the meaning of `content`.
If you want to use an article, I think "the" is more appropriate because it refers to the specific log message that is passed as the `content` parameter.

kangz12345 replied:
I think these are quite subtle and you only have to be careful when you write sentences.

Aijuh replied:
I will go back to previous one.

Aijuh committed:
Change comments.

kangz12345 reviewed (approved):
LGTM

BECATRUE reviewed (approved):
LGTM!
